### ISSUE TITLE
why result to the train result??

### ISSUE THREAD

**Issue by mcjqwer** [ultralytics/yolov5#5398]
## ❔Question
I train the yolov5 3.1version. 1500epoch  , m weight , batchsize 64, 
and the result looks good. But the result.png as blow.
 ## Additional context

**Comment by mcjqwer** [comment]
In particular the curve of val object、 P、R、map0.5:0.95

**Comment by glenn-jocher** [comment]
@mcjqwer 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by mcjqwer** [comment]
> @mcjqwer 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results.
> 
> Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.
> 
> If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.
> 
> We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.
> 
> ## Dataset
> * **Images per class.** ≥ 1500 images per class recommended
> * **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
> * **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
> * **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
> * **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
> * **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.
> 
> [<img alt="COCO Analysis" width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png">](https://arxiv.org/abs/1405.0312)
> 
> ## Model Selection
> Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models.
> 
> <img alt="YOLOv5 Models" width="700" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png">
> 
> * **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. VOC, VisDrone, GlobalWheat). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
> 
> ```shell
> python train.py --data custom.yaml --weights yolov5s.pt
>                                              yolov5m.pt
>                                              yolov5l.pt
>                                              yolov5x.pt
>                                              custom_pretrained.pt
> ```
> 
> * **Start from Scratch.** Recommended for large datasets (i.e. COCO, Objects365, OIv6). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
> 
> ```shell
> python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
>                                                       yolov5m.yaml
>                                                       yolov5l.yaml
>                                                       yolov5x.yaml
> ```
> 
> ## Training Settings
> Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.
> 
> * **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
> * **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
> * **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
> * **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).
> 
> ## Further Reading
> If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Thank you for your reply , i don't do any change to the orign model  , i train the person head and  shoulder as my object. 
This is my exp file and part of my dataset(the whole dataset has 3200 image and 40000 object). I will be preasure if you can read me !
![labels](https://user-images.githubusercontent.com/59089548/139444179-c6a4d291-f907-4911-adf2-ddd9313121bc.png)
![labels_correlogram](https://user-images.githubusercontent.com/59089548/139444197-b2f9f4b1-ab23-498d-986c-303f8d83b0b8.png)
![precision-recall_curve](https://user-images.githubusercontent.com/59089548/139444200-7cbeaddf-2c5e-49af-b7c3-20946590141c.png)
![results](https://user-images.githubusercontent.com/59089548/139444202-b7e352d6-76e7-4851-960a-e2c0e36c2cb2.png)
[results.txt](https://github.com/ultralytics/yolov5/files/7442287/results.txt)
![test_batch0_gt](https://user-images.githubusercontent.com/59089548/139444216-c466147d-124a-4dc1-b166-26731345103b.jpg)
![test_batch0_pred](https://user-images.githubusercontent.com/59089548/139444221-ded4afa2-098f-4218-9a55-c6e9d3a4e866.jpg)
![train_batch0](https://user-images.githubusercontent.com/59089548/139444223-fa48ac27-f8fa-4a2f-86a7-b5cc9bff735b.jpg)
![train_batch1](https://user-images.githubusercontent.com/59089548/139444226-9d4a16a3-9a6c-4f38-96c8-3ab8f1248290.jpg)
![train_batch2](https://user-images.githubusercontent.com/59089548/139444228-a1022b39-3f77-4488-9179-b74fe2607d42.jpg)
![0](https://user-images.githubusercontent.com/59089548/139444273-c437bfea-b223-425c-baca-36045e7457db.jpg)
![20](https://user-images.githubusercontent.com/59089548/139444282-cb43dd07-82f4-4ae0-af07-814ee5a30ba8.jpg)
![40](https://user-images.githubusercontent.com/59089548/139444287-c218eee1-5ede-45f8-92d9-d6ea17833a1a.jpg)
![160](https://user-images.githubusercontent.com/59089548/139444291-c247ee0b-aa31-435a-905f-37f8c3854c7a.jpg)
![180](https://user-images.githubusercontent.com/59089548/139444293-70952850-3216-4ddc-a16a-68a22b9e9ade.jpg)
![200](https://user-images.githubusercontent.com/59089548/139444294-219c2395-3e7a-4b90-9c8b-e4d294f7b31f.jpg)
![320](https://user-images.githubusercontent.com/59089548/139444299-b817b2fb-a928-49d7-8e74-af65e774d9bd.jpg)
![340](https://user-images.githubusercontent.com/59089548/139444303-5e5ce67d-05ab-4b44-9cf2-5540b8b575a1.jpg)
![360](https://user-images.githubusercontent.com/59089548/139444307-c60e59bc-3029-4581-9ddb-f45bc03b525b.jpg)

**Comment by glenn-jocher** [comment]
@mcjqwer got it! Everything looks good, except that you've trained too long. Your model is overfitting past 100 epochs, so you should reduce --epochs to 100 and maybe experiment with various levels of augmentation, i.e.:
```
python train.py --epochs 100 --hyp hyp.scratch-low.yaml
python train.py --epochs 100 --hyp hyp.scratch-med.yaml
python train.py --epochs 100 --hyp hyp.scratch-high.yaml
```

EDIT: note that even though you've overtrained, your best.pt will still be good, it was last saved around epoch 100 and always has your best results. last.pt is at epoch 1000 and will show worse results.
